bindings/go/cel/expression/parser/cel.go

```diff
@@ -0,0 +1,124 @@
+package parser
+
+import (
+	"errors"
+	"strings"
+)
+
+const (
+	// CEL expressions are enclosed between "${" and "}"
+	exprStart = "${"
+	exprEnd   = "}"
+)
+
+// ErrNestedExpression is returned when an expression contains another
+// unescaped expression. Nested expressions are only allowed when the
+// inner expression appears inside a string literal, for example:
+//
+//	${outer("${inner}")}   // allowed
+//	${outer(${inner})}     // not allowed
+var ErrNestedExpression = errors.New("nested expressions are not allowed unless inside string literals")
+
+// extractExpressions extracts all non-nested CEL expressions from a string.
+// It returns an error if it encounters an illegal nested expression.
+func extractExpressions(str string) ([]string, error) {
```

Comment:
This is a nit however, just want to put this out there. I would love for this to be on a struct with proper internal indexes keeping track for where they are and we could call things like `peekChar` and `skipWhitespace`. It would read so much better. And we always write code for reading, and not writing.

Hopefully, we write this once and then never look at it again, but I guarantee it that we will have to do that eventually. :)  For now, this isn't that complex to understand. Yet. :)